grant-project/zcash-grant-system
# Failed proposal refunds

Most of #99. Closes #233.

### What This Does

* Adds a job type that, when a proposal is published, will check if it failed to reach its funding goal. This puts it into the REFUNDING stage. This emails all contributors and the proposal team.
* Adds an alert and admin action for attaching a refund txid to a contribution. This emails the contributor. There's also a homepage callout and filter for refundable contributions.
* Adds a refund_address to user settings that they can configure.

One of the open questions is the proposal stake. Right now it's marked as refundable, but should this be the case?

### What This Doesn't Do

* Give admins a way to refund successful projects (e.g. in case of being abandoned)
* Actually make jobs run regularly

### Steps to Test

* Create a proposal via auto-fill so that it gets the 5 minute deadline
* Have a user partially fund the project, but don't push it over
* After 5 minutes, hit the endpoint `localhost:5000/api/v1/task`, confirm that the task completes, the proposal is put into REFUNDING stage, and everyone involved was emailed about the failure
* Check that the admin doesn't show any refundable contributions
* Go to the contribution user's settings and set a refund address
* Check the admin again, confirm it notifies you of the refundable tx
* Go to the contribution, confirm it gives you a notice to fill out a txid
* Enter anything, confirm it marks it as refunded, emails you, and stops alerting admins to it

## Screenshots

<img width="591" alt="screen shot 2019-02-19 at 12 27 53 pm" src="https://user-images.githubusercontent.com/649992/53035042-87107780-3442-11e9-9ad6-8ad30ba5cf8f.png">

<img width="655" alt="screen shot 2019-02-19 at 12 27 58 pm" src="https://user-images.githubusercontent.com/649992/53035049-8972d180-3442-11e9-829d-5c949c889157.png">

<img width="923" alt="screen shot 2019-02-19 at 12 28 03 pm" src="https://user-images.githubusercontent.com/649992/53035058-8d9eef00-3442-11e9-8814-8f31dfb2a81e.png">

<img width="863" alt="screen shot 2019-02-19 at 12 31 29 pm" src="https://user-images.githubusercontent.com/649992/53035062-9099df80-3442-11e9-9ab5-07525663968b.png">
